Helper/Discount.php

```diff
@@ -0,0 +1,320 @@
+     * @param \Amasty\GiftCard\Model\Account $accountModel
+     * @param Quote $quote
+     * @return float
+     * @throws LocalizedException
+     */
+    public function applyAmastyGiftCard($code, $accountModel, $quote)
+    {
+        // Get current gift card balance before applying it to the quote
+        // in case "fixed_amount" / "pay for everything" discount type is used
+        $giftAmount = $accountModel->getCurrentValue();
+
+        $quoteId = $quote->getId();
+
+        $isValid = $this->amastyGiftCardManagement->getInstance()->validateCode($quote, $code);
+
+        if (!$isValid) {
+            throw new LocalizedException(__('Coupon with specified code "%1" is not valid.', $code));
+        }
+
+        if ($accountModel->canApplyCardForQuote($quote)) {
+            $quoteGiftCard = $this->amastyQuoteFactory->getInstance()->create();
+            $this->amastyQuoteResource->getInstance()->load($quoteGiftCard, $quoteId, 'quote_id');
+            $subtotal = $quoteGiftCard->getSubtotal($quote);
+
+            if ($quoteGiftCard->getCodeId() && $accountModel->getCodeId() == $quoteGiftCard->getCodeId()) {
+
+                throw new LocalizedException(__('This gift card account is already in the quote.'));
+
+            } elseif ($quoteGiftCard->getGiftAmount() && $subtotal == $quoteGiftCard->getGiftAmount()) {
+
+                throw new LocalizedException(__('Gift card can\'t be applied. Maximum discount reached.'));
+
+            } else {
+                $quoteGiftCard->unsetData($quoteGiftCard->getIdFieldName());
+                $quoteGiftCard->setQuoteId($quoteId);
+                $quoteGiftCard->setCodeId($accountModel->getCodeId());
+                $quoteGiftCard->setAccountId($accountModel->getId());
+
+                $this->amastyQuoteRepository->getInstance()->save($quoteGiftCard);
+                $this->updateTotals($quote);
+
+                if ($this->getAmastyPayForEverything()) {
+                    // pay for everything, items, shipping, tax
+                    return $giftAmount;
+                } else {
+                    // pay for items only
+                    $totals = $quote->getTotals();
+                    return $totals[self::AMASTY_GIFTCARD]->getValue();
+                }
+            }
+        } else {
+            throw new LocalizedException(__('Gift card can\'t be applied.'));
+        }
+    }
+
+    /**
+     * Try to copy Amasty Gift Cart data from parent to immutable quote
+     *
+     * @param int|string $parentQuoteId
+     * @param int|string $immutableQuoteId
+     */
+    public function cloneAmastyGiftCards($parentQuoteId, $immutableQuoteId)
+    {
+        try {
+            $connection = $this->resource->getConnection();
+            $giftCardTable = $this->resource->getTableName('amasty_amgiftcard_quote');
+
+            // Clear previously applied gift cart codes from the immutable quote
+            $sql = "DELETE FROM {$giftCardTable} WHERE quote_id = :quote_id";
```

Comment:
These should be in some sort of transaction, so we can roll back in case of disaster. What if we do a successful `DELETE` but then our connection is interrupted before we do our `INSERT`? That applies to most of the database interactions throughout.

Reply:
@danac-gs Any plans for wrapping this in a transaction? Or is this going to be called somewhere else where we can roll it back?

Reply:
@matt-thomason It is wrapped. Not sure why you don't see it. Here is the link to the file:
https://github.com/BoltApp/bolt-magento2/blob/f7949f38f77cc3c9a5160cd7b3ec50363aa3b896/Helper/Discount.php#L232

Reply:
huh I'm crazy